### ISSUE TITLE
CXF plugin didn't work

### ISSUE THREAD

**Issue by koala775** [naver/pinpoint#4596]
**What version of pinpoint are you using?**
v1.7.3.

**Describe your problem**
I have two jboss application A and B . A invoke B by cxf , but there is no data collector and no trace shows up in the web UI .

**What have you done?**
Steps you have taken to trigger the problem, for example:
1. Launched JBOSS with agent respectively.
2. Tried tracing 'cxf plugin'
3. No trace shows up in the web UI and as follows ,server map option is setted: 
![image](https://user-images.githubusercontent.com/35438579/45194170-7dc09380-b284-11e8-8552-60eb56e89bfd.png)

**Additional context**
Add any other context about the problem here, such as affected library for agents, how your collector/web/hbase is set up if applicable.
collector:default
web:default
agent:pinpoint.config
profiler.cxf.client=true
profiler.cxf.client.hiddenParams=false
hbase:hbase-site.xml
<configuration>
 <property>
    <name>hbase.rootdir</name>
    <value>file:///opt/applog/hbase/data</value>
 </property>
 <property>
    <name>hbase.zookeeper.property.dataDir</name>
    <value>/opt/applog/zookeeper/data</value>
 </property>
 <property>
    <name>hbase.master.port</name>
    <value>60000</value>
 </property>
 <property>
    <name>hbase.regionserver.port</name>
    <value>60020</value>
 </property>
</configuration>

**Comment by koala775** [comment]
Hi @Xylus 
agent DEBUG log is below:
2018-09-07 10:44:22 [DEBUG](c.n.p.p.i.JavassistMethod          ) addBeforeInterceptor before behavior:org.apache.cxf.frontend.ClientProxy.invokeSync(java.lang.reflect.Method,org.apache.cxf.service.model.BindingOperationInfo,java.lang.Object[]) code:{try { _$PINPOINT$_interceptor661 = com.navercorp.pinpoint.bootstrap.interceptor.registry.InterceptorRegistry.getInterceptor(661); ((com.navercorp.pinpoint.bootstrap.interceptor.AroundInterceptor)_$PINPOINT$_interceptor661).before(this, $args);} catch (java.lang.Throwable _$PINPOINT_EXCEPTION$_) { com.navercorp.pinpoint.bootstrap.interceptor.InterceptorInvokerHelper.handleException(_$PINPOINT_EXCEPTION$_); }}

2018-09-07 10:44:22 [DEBUG](c.n.p.p.i.JavassistMethod          ) addAfterInterceptor after behavior:org.apache.cxf.frontend.ClientProxy.invokeSync(java.lang.reflect.Method,org.apache.cxf.service.model.BindingOperationInfo,java.lang.Object[]) code:{try { ((com.navercorp.pinpoint.bootstrap.interceptor.AroundInterceptor)_$PINPOINT$_interceptor661).after(this, $args, ($w)$_, null);} catch (java.lang.Throwable _$PINPOINT_EXCEPTION$_) { com.navercorp.pinpoint.bootstrap.interceptor.InterceptorInvokerHelper.handleException(_$PINPOINT_EXCEPTION$_); }}

**Comment by Xylus** [comment]
@koala775 Are the `before()` and `after()` methods from **CxfClientInvokeSyncMethodInterceptor** actually called?

**Comment by Xylus** [comment]
The logs above are from class load time, saying that the interceptors are added to **ClientProxy**'s *invokeSync* method.
You'll have to look at the log of when the cxf call is made and see if **CxfClientInvokeSyncMethodInterceptor** has left a log in it's *before* and *after* methods.

**Comment by koala775** [comment]
@Xylus
 CxfClientInvokeSyncMethodInterceptor appear in below:
![image](https://user-images.githubusercontent.com/35438579/45198797-f7637c00-b29a-11e8-9138-154acb817bb3.png)

and is there to see?

**Comment by Xylus** [comment]
@koala775 Those are from class load time (where class file transformation is happening).
What you're looking for is something like below :
```
... BEFORE org.apache.cxf.frontend.ClientProxy  args:(...)

... AFTER org.apache.cxf.frontend.ClientProxy  args:(...)
```

**Comment by koala775** [comment]
@Xylus 
I didn't see what you said in the log.

... BEFORE org.apache.cxf.frontend.ClientProxy  args:(...)
... AFTER org.apache.cxf.frontend.ClientProxy  args:(...)

**Comment by Xylus** [comment]
Hmm, then I don't think the interceptors were called.
Could you post how you're making cxf calls?
There might be cases where the cxf plugin doesn't trace, for example if cxf calls are made by differently as to not invoke `ClientProxy.invokeSync(...)`.
If this is the case, then support for tracing would need to be added to the plugin.

**Comment by koala775** [comment]
@Xylus 
as below:
![image](https://user-images.githubusercontent.com/35438579/45200238-727c6080-b2a2-11e8-9622-b9f2bbc42f1a.png)
![image](https://user-images.githubusercontent.com/35438579/45200283-ac4d6700-b2a2-11e8-92cf-269087d60883.png)

**Comment by Xylus** [comment]
Thanks, I'll take a look and see what's happening.

**Comment by koala775** [comment]
@Xylus 
supplement:
![image](https://user-images.githubusercontent.com/35438579/45200922-a1480600-b2a5-11e8-9878-a42d1fdad259.png)

**Comment by koala775** [comment]
@Xylus 
I use org.apache.cxf.jaxws.JaxWsClientProxy.invoke method，cxf version is 2.2.8 .
thank you for you help!

**Comment by koala775** [comment]
@Xylus
Hi ， I have rewrited cxf and I have see below log ，but  no trace shows up in the web UI
2018-09-11 09:15:29 [DEBUG](eefareClientInvokeMethodInterceptor) BEFORE org.apache.cxf.jaxws.JaxWsClientProxy  args:($Proxy99, Method, Object[])
2018-09-11 09:15:29 [DEBUG](eefareClientInvokeMethodInterceptor) BEFORE org.apache.cxf.jaxws.JaxWsClientProxy  args:($Proxy99, Method, Object[])

2018-09-11 09:15:50 [DEBUG](eefareClientInvokeMethodInterceptor) AFTER org.apache.cxf.jaxws.JaxWsClientProxy  args:($Proxy99, Method, Object[])
2018-09-11 09:15:50 [DEBUG](eefareClientInvokeMethodInterceptor) AFTER org.apache.cxf.jaxws.JaxWsClientProxy  args:($Proxy99, Method, Object[])

**Comment by Xylus** [comment]
@koala775 
The cxf plugin currently seems to simply record the `JaxWsClientProxy.invokeSync(...)` method and not propagate trace over the network.
So the trace will only show up as a method invocation in the call stack at the transaction view, and won't render a remote node in the server map.

I think the cxf plugin needs [this](http://naver.github.io/pinpoint/plugindevguide.html#231-if-the-next-node-is-traceable)(client side) and [this](http://naver.github.io/pinpoint/plugindevguide.html#22-top-level-method-of-a-node)(server side) implemented to get what you're looking for.
(Please feel free to implement it and throw us a PR if you'd like :))

**Comment by koala775** [comment]
@Xylus 
So, I need write two plugins ,cxf_client and cxf_server?

**Comment by Xylus** [comment]
You can probably have them all in a single plugin. Better yet, you can probably add the implementation into the existing cxf plugin.

**Comment by koala775** [comment]
@Xylus 
Can I pass data to the next node by cxf ,like http as follows:
![image](https://user-images.githubusercontent.com/35438579/45338453-e333bd80-b5bf-11e8-88bd-0693c26e3ab5.png)

**Comment by Xylus** [comment]
Yes, that is exactly what is required to propagate trace over to a remote node. Good job!
I don't know what protocol cxf uses for rpc, but if there is a header space  one way of doing it would be to pass those data inside the header just like how trace is propagated over HTTP headers.

**Comment by koala775** [comment]
@Xylus 
How can I have client and server all in a single plugin? What should I do in before and after method?

**Comment by koala775** [comment]
@Xylus 
I saw official website  pinpoint-plugin-sample project, there is a TargetClass13_Request class and this can use to pass data directly?like this:
![image](https://user-images.githubusercontent.com/35438579/45341072-bc7b8400-b5cb-11e8-8aa7-14bb1037f77c.png)

**Comment by koala775** [comment]
@Xylus 
The IcePlugin is write by myself , but hava error :
![image](https://user-images.githubusercontent.com/35438579/45477047-313cf280-b773-11e8-95b8-e3256b2c7df5.png)

**Comment by VictorZeng** [comment]
Which version of CXF?

**Comment by Xylus** [comment]
Possibly resolved via #4701
